Make `torch.fx` compatible with Python-3.11

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92787
* __->__ #92895

In 3.11 bytecode size is not constant, so in order to get from `f_lasti` to opcode index, one need to search for the closes offset in disassembled instructions.

Update `_patch_function` to construct code with all the properties that exist in 3.11 runtime.
Update `_torchscript_schema_to_signature` to mark `from` named arg as positional argument only, as this is a reserved keyword in Python and as such checked by `inspect` package in 3.11

malfet committed:
Make `torch.fx` compatible with Python-3.11

In 3.11 bytecode size is not constant, so in order to get from lasti to
opcode one need to search for the one in disassembled instructions
offset

[ghstack-poisoned]

malfet committed:
Update on "Make `torch.fx` compatible with Python-3.11"


In 3.11 bytecode size is not constant, so in order to get from lasti to
opcode one need to search for the one in disassembled instructions
offset

[ghstack-poisoned]

albanD commented on torch/fx/proxy.py:
```diff
@@ -358,7 +359,13 @@ def __iter__(self) -> Iterable['Proxy']:
         assert calling_frame is not None
-        inst = list(dis.get_instructions(calling_frame.f_code))[calling_frame.f_lasti // 2]
+        inst_list = list(dis.get_instructions(calling_frame.f_code))
+        if sys.version_info >= (3, 11):
```
This might need to become a helper if we need it in more places.

albanD commented on torch/fx/_symbolic_trace.py:
```diff
@@ -119,7 +119,28 @@ def _patch_function(fn: FunctionType, nargs: int) -> FunctionType:
     co_flags = co.co_flags & ~HAS_VARSTUFF
     co_args: tuple
-    if hasattr(co, "co_posonlyargcount"):
+    if hasattr(co, "co_qualname"):
```
I guess this needs to be guarded for 3.11 to make the lint happy?

malfet replied:
Not sure if `mypy` is smart enough to detect it, but added ignore to the two violations

albanD reviewed (approved):
The fix sounds good, just a bit of massage for the lint.

malfet committed:
Update on "Make `torch.fx` compatible with Python-3.11"


In 3.11 bytecode size is not constant, so in order to get from `f_lasti` to opcode index, one need to search for the closes offset in disassembled instructions.

Update `_patch_function` to construct code with all the properties that exist in 3.11 runtime


[ghstack-poisoned]

malfet committed:
Update on "Make `torch.fx` compatible with Python-3.11"


In 3.11 bytecode size is not constant, so in order to get from `f_lasti` to opcode index, one need to search for the closes offset in disassembled instructions.

Update `_patch_function` to construct code with all the properties that exist in 3.11 runtime


[ghstack-poisoned]

albanD reviewed (approved):
Sounds ok to me

malfet commented:
@pytorchbot merge

malfet commented:
Windows failed with `RuntimeError: CUDA error: the provided PTX was compiled with an unsupported toolchain.
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.`, which is highly unlikely to have caused by this change

malfet commented:
@pytorchbot merge -f "Windows failure is flake"
